[CONTENT] Blog: GUI setup, Part 1 (#123)
reactos/web-content: content/blogs/gui-setup-part1-setupapi.md

---
title:      "1st-stage GUI setup - September 2023: partly Wine-syncing setupapi"
author:     "hbelusca"
date:       2023-10-30
tags:       [ newsletter, gui-setup ]
---

Greetings to all ReactOS followers!

> As many of you certainly are aware by now, [I have been officially hired](/project-news/hermes-belusca-hired-full-time) by the ReactOS Deutschland e.V. to develop the graphical version of the 1st-stage ReactOS installer ("1st-stage GUI setup").

Author: Question for maybe @ColinFinck or others:
is it OK to say "hired by the ReactOS Deutschland e.V." ? or maybe simpler, "hired by the ReactOS project" ?

During this first month (September 2023), my goal was to partly sync the code of the `setupapi.dll` module with Wine's.
Indeed, this module was [forked](https://git.reactos.org/?p=reactos.git;a=blob;f=media/doc/WINESYNC.txt;h=f6614ac83a968cd5d7ddde6483dc85b94d264fc4;hb=HEAD#l182) way back in 2005, with later partial synchronizations of its code for some of its functions.

For those of you who do not know what [setupapi](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/setupapi) is, it is a module that:

> - exposes [generic setup functions](https://learn.microsoft.com/en-us/windows/win32/setupapi/setup-api-portal) for reading and processing INF files, moving/copying files from an installation source media to a target..., supporting also extraction from compressed .CAB cabinet files;

Reviewer: ellipsis and comma next to eachother.

(NOTE: It should be observed that our current text-mode 1st-stage installer, USETUP, re-implements a very small subset of these functions, used for extracting files from the `reactos.cab` source file and copying them to the ReactOS installation directory.
In the GUI setup, my aim is to use directly those exposed by `setupapi.dll`.)

The different sub-tasks I need to address are listed in this JIRA report [CORE-19209](https://jira.reactos.org/browse/CORE-19209).

## Winesync.py script

This endeavour started by trying to use our in-house `winesync.py` python script to help syncing Wine code.
It requires one to write a configuration file (e.g. `setupapi.cfg`) that lists which files/directories can be synced for the given module, and maps these from their names in the [Wine repository](https://github.com/wine-mirror/wine/tree/master/), for example: [`dlls/setupapi`](https://github.com/wine-mirror/wine/tree/master/dlls/setupapi) (or one or more of its files), to their corresponding path names in the [ReactOS repository](https://github.com/reactos/reactos/tree/master/dll/win32/setupapi).

> As it is customary with any ReactOS developing exercise, one always encounters problems or bugs, and my attempt at using the `winesync.py` script was no exception to this rule.

Reviewer: ```suggestion
With any ReactOS development task, one always encounters problems or bugs. My attempt at using the `winesync.py` script was no exception to this rule.
```

This sentence has more then one comma and it's a little hard to understand, this is my interpretation

The script supposes that one always syncs code with both the Wine and the [Wine-staging](https://github.com/wine-staging/wine-staging) repositories; the script does not support doing the sync with just the Wine repository alone. This had to be fixed.

Next, I develop ReactOS on my old laptop computer running Windows 7, using [Git-for-Windows](https://gitforwindows.org/) and python there.
The `winesync.py` script expected its user to run it under a Unix-like environment, and supposed at many places that the path-manipulation operations were always done in such context.
When run under Python on Windows, these operations generated Windows-formatted paths, that were not always understood (by the pygit module, mainly).
(Note that due to my setup I cannot use WSL there, as suggested by some!)

Other similar problems had to be addressed as well, and I will not list them all there. You can consult the list of required fixes and improvements in the following [PR #5898](https://github.com/reactos/reactos/pull/5898) (NOTE: it has not been merged into the master branch at the time of writing).

## (Partly) wine-syncing setupapi proper

> After these fixes in the syncing script, I started the partial syncing of the setupapi module.

Reviewer: ```suggestion
After these fixes in the winesync script, I started the partial syncing of the setupapi module.
```

---
title:      "1st-stage GUI setup - September 2023: partly Wine-syncing setupapi"
author:     "hbelusca"
date:       2023-10-30
tags:       [ newsletter, gui-setup ]
---

Greetings to all ReactOS followers!

> As many of you certainly know by now, I have been officially hired by the ReactOS Deutschland e.V. to develop the graphical version of the 1st-stage ReactOS installer ("1st-stage GUI setup").

Reviewer 2: I think it would be better to elaborate a bit more on this, or link to another blog post explaining it in detail.

Author: Will do

Reviewer 2: works for me.

During this first month (September 2023), my goal was to partly sync the code of the setupapi.dll module with Wine's.
Indeed, this module was [forked](https://git.reactos.org/?p=reactos.git;a=blob;f=media/doc/WINESYNC.txt;h=f6614ac83a968cd5d7ddde6483dc85b94d264fc4;hb=HEAD#l182) way back in 2005, with later partial synchronizations of its code for some of its functions.

> For those of you who don't know what [setupapi](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/setupapi) is, it is a module that: (i) exposes [generic setup functions](https://learn.microsoft.com/en-us/windows/win32/setupapi/setup-api-portal) for reading and processing INF files, moving/copying files from an installation source media to a target..., supporting also extraction from compressed .CAB cabinet files; and (ii) exposes device installation functions as well.

Reviewer 2: I would add bullet points here, instead of (i), (ii).

Because of its very nature, it is "one of those Wine components we don't share completely because of architectural differences" [(CORE-7662)](https://jira.reactos.org/browse/CORE-7662) (due mainly to the device installation functions), and as such, as time goes by, Wine code improvements for code that isn't tied to OS- and architecture-specific aspects might go unnoticed and our shared code slowly drifts apart for no good reason.

As part of the 1st-stage GUI setup, I need to use the [File Queues](https://learn.microsoft.com/en-us/windows/win32/setupapi/file-queues) and [Cabinet Files](https://learn.microsoft.com/en-us/windows/win32/setupapi/cabinet-files--setup-api-) functions. These are those that have received recent improvements/bug-fixes from Wine, and that can be synchronized without too much trouble.

> (NOTE: It should be observed that our current text-mode 1st-stage installer, USETUP, re-implements a very small subset of these functions, used for extracting files from the reactos.cab source file and copying them to the ReactOS installation directory. In the GUI setup, my aim is to use directly those exposed by setupapi.dll.)

Reviewer 2: Starting a new line after each full-stop makes it easier to leave comments per-line

> (NOTE: It should be observed that our current text-mode 1st-stage installer, USETUP, re-implements a very small subset of these functions, used for extracting files from the reactos.cab source file and copying them to the ReactOS installation directory. In the GUI setup, my aim is to use directly those exposed by setupapi.dll.)

Reviewer 2: Maybe try putting single backticks around `reactos.cab`, see if that highlights the single word?

The different sub-tasks I need to address are listed in this JIRA report [CORE-19209](https://jira.reactos.org/browse/CORE-19209).

## Winesync.py script

> This endeavour beginned by trying to use our in-house `winesync.py` python script to help syncing Wine code. It requires one to write a configuration file (e.g. `setupapi.cfg`) that lists which files/directories can be synced for the given module, and maps these from their names in the [Wine repository](https://github.com/wine-mirror/wine/tree/master/), for example: [`dlls/setupapi`](https://github.com/wine-mirror/wine/tree/master/dlls/setupapi) (or one or more of its files), to their corresponding path names in the [ReactOS repository](https://github.com/reactos/reactos/tree/master/dll/win32/setupapi).

Reviewer 2: `s/beginned/started`

As it is customary with any ReactOS developing exercise, one always encounters problems or bugs, and my attempt at using the `winesync.py` script was no exception to this rule.
The script supposes that one always syncs code with both the Wine and the [Wine-staging](https://github.com/wine-staging/wine-staging) repositories; the script does not support doing the sync with just the Wine repository alone. This had to be fixed.

> Next, I develop ReactOS on Windows, and so, I use Git4Windows and python there. The `winesync.py` script expected its user to run it under a Unix-like environment, and supposed at many places that the path-manipulation operations were always done in such context. When run under Python on Windows, these operations generated Windows-formatted paths, that were not always understood (by the pygit module, mainly).

Reviewer 2: Could try wsl?

Author: Comment about that added :P
